XMonad/Hooks/DynamicLog.hs

```diff
@@ -585,6 +589,100 @@ pprWindowSetXinerama ws = "[" ++ unwords onscreen ++ "] " ++ unwords offscreen
         offscreen = map S.tag . filter (isJust . S.stack)
                         . sortOn S.tag $ S.hidden ws
 
+
+-- | Custom Datatype for custom icons based on the state of the 'Workspace'
+data NameSet l = NameSet 
+    { nameSetCurrent :: l -- ^ If the 'Workspace' is the current workspace
+    , nameSetVisible :: l -- ^ If the 'Workspace' is visible (Xinerama only)
+    , nameSetHidden :: l -- ^ If the 'Workspace' isnt visible but still has windows
+    , nameSetHiddenNoWindows :: l -- ^ If the 'Workspace' isnt visible and has no windows
+    , showOverlayIcons :: Bool  -- ^ An option that would be use if this was integrated into a set of `Icon` for base 'Workspace's
+    }
+type Icon = NameSet String -- ^ Type alias to make 'NameSet' specific to Strings
```

Comment:
1. This could use some strictness annotations; at least I don't see why these fields should not be strict.
2. Why is the `NameSet` defined with a type variable here?  Everything is explicitly specialized to `Icon` throughout the module and you're not deriving anything that would need a `Type -> Type` kind; is there some use-case you have in mind for this?
3. This could use some documentation and usage examples.